metadatastore/commands.py

```diff
@@ -39,88 +89,273 @@ def inner(*args, **kwargs):
+
+    # query the database for any event descriptors which
+    # refer to the given runstart
+    ev_desc_cur = EventDescriptor._get_collection().find(
+        {'run_start_id': oid})
+
+    # loop over the found documents, cache, and dereference
+    rets = [_cache_eventdescriptor(ev_desc) for ev_desc in ev_desc_cur]
+
+    # if nothing found, raise
+    if not rets:
+        raise ValueError("No EventDescriptors exists")
+
+    # return the list of event descriptors
+    return rets
+
+
+def fetch_events_generator(desc_uid):
+
+    col = Event._get_collection()
+
+    desc = event_desc_given_uid(desc_uid)
+    eid = _EVENTDESC_UID_to_OID_MAP[desc_uid]
+
+    ev_cur = col.find({'descriptor_id': eid},
+                      sort=[('time', 1)])
+
+    for ev in ev_cur:
+        # ditch the ObjectID
+        ev.pop('_id')
+        # pop the descriptor oid
+        ev.pop('descriptor_id')
+        # replace it with the defererenced descriptor
+        ev['descriptor'] = desc
+        # pop the data
+        data = ev.pop('data')
+        # replace it with the friendly paired dicts
+        ev['data'], ev['timestamps'] = [{k: v[j] for k, v in data.items()}
+                                        for j in range(2)]
+        # wrap it our fancy dict
+        ev = doc.Document('Event', ev)
+
+        yield ev
+
+
+def fetch_events_table(descriptor):
+    """Return all event data as tables
+
+    Parameters
+    ----------
+    descriptor : dict or uid
+        The EeventDestriptor to get the events for.  Can be either
+        a dict or a uid.
+
+    Returns
+    -------
+    descriptor : dict
+        EventDescriptor document
+
+    data_table : DataFrame
+        All of the data as a DataFrame indexed on sequence number
+
+    uids : Series
+        The uids of each of the events as a `Series` indexed on sequence number
+
+    times : Series
+        The times of each event as a `Series` indexed on sequence number
+
+    uids : Series
+        The uids of each event as a `Series` indexed on sequence number
+
+    timestamps_table : DataFrame
+        The timestamps of each of the measurements as a `DataFrame`.  Same
+        columns as `data_table` indexed on sequence number
+    """
+    import pandas as pd
+    desc_uid = doc_or_uid_to_uid(descriptor)
+    descriptor = event_desc_given_uid(desc_uid)
+    # this will get more complicated once transpose caching layer is in place
+    all_events = list(fetch_events_generator(desc_uid))
+
+    # get event sequence numbers
+    seq_num = pd.Series([ev['seq_num'] for ev in all_events])
+
+    # get event times
+    times = pd.to_datetime([ev['time'] for ev in all_events], unit='s')
+    times.index = seq_num
+
+    # get uids
+    uids = pd.Series([ev['uid'] for ev in all_events], index=seq_num)
+
+    # get data values
+    data_table = pd.DataFrame([ev['data'] for ev in all_events], index=seq_num)
+
+    # get timestamps
+    ts = pd.DataFrame([ev['timestamps'] for ev in all_events], index=seq_num)
```

Comment:
Iterating over `all_events` five times feels a little dirty. Wouldn't it be faster to iterate over it once and build these five lists in the same for loop? Maybe use a deque?